sdk/core/azure-core/azure/core/credentials.py

```diff
@@ -92,6 +93,70 @@ def update(self, key: str) -> None:
         self._key = key
 
 
+class AzureDuoKeyCredential:
```

Comment:
Why not just take a tuple/set of keys in the existing AzureKeyCredential/support multiple headers/keys in the existing policy?

E.g. something like:

```python
import typing

TKeys = typing.TypeVar('TKeys', bound = str | typing.Tuple[str, ...])

class KeyCredential(typing.Generic[TKeys]):

    def __init__(self, keys: TKeys):
        self._keys = keys

    @property
    def keys(self) -> typing.Sequence[str]:
        return self._keys


class ClientThatNeedsMultiKeyCredentials:

    def __init__(self, *, credentials: KeyCredential[typing.Tuple[str, str]]):
        self._credentials = credentials


c = ClientThatNeedsMultiKeyCredentials(credentials=KeyCredential(keys=('one', 'two')))
```

Reply:
+ @kristapratico , @annatisch to get their thoughts as well. Typing is fun!

Reply:
Add ctor signatures here to make it easier to read.

Current ctor of AzureKeyCredential is 

```py
def __init__(self, key: str) -> None:
```

And ctor of AzureKeyCredentialPolicy is
```py
def __init__(
        self,
        credential: "AzureKeyCredential",
        name: str,
        **kwargs  # pylint: disable=unused-argument
    ) -> None:
```

Reply:
This is my thought but any feedback is welcome. :)

Introducing new credential/policy vs updating existing AzureKeyCredential/AzureKeyCredentialPolicy. Currently in the AzureKeyCredential/AzureKeyCredentialPolicy, users store the API key in the credential and the header in the policy. In the two header-api key scenarios, we need to store the header and corresponding api key in one place (we also need to store the mapping information). Hence, we need to make non-trivial changes on the surfaces. E.g. the policy ctor does not require the header any more, credential.update() needs the header in addition to the value, etc.. Which makes me think it is better and simpler to introduce a new one.

Reply:
I like it, but should it be a union of `str | typing.Dict[str, str]`? It looks like a header targets a particular key. @xiangyan99 correct me if I'm wrong about that

Reply:
@kristapratico yes. I think we need the header: key mapping information.

Reply:
We either rely on the policy to "own" the header names (and have a prescribed/documented order) or we do use a mapping (no need for it to be mutable :)). The existing (single) key credential does not take the header name as a parameter, so I didn't expose it to the end-user of the SDK either. I'd probably argue that the generic (no pun intended) names for the parameters are not moving the needle/still require the caller/user of the SDK to know the mapping much in the same way as they would if the API required a tuple.

I'm pretty sure one can `typing.Generic`ify the interdependency between the number of parameters to the credential and consuming policy if one tries hard enough. Naming the parameters is more interesting. Maybe use `typing.TypedDict`? :)

Reply:
Under what circumstances do we believe that it is beneficial (or required) for the user instantiating the credential to know the exact header values?

Reply:
(or, to put it another way, when does the caller know better than the client which header names to use)

Reply:
Ok - one observation is that this leaks the `header1` and `header2` property names to the user of the SDK as well as the header names in the `update` method if they want to roll keys.

I don't like type proliferation - especially for things like credentials. If this is effectively a protocol/abstract base class (as in no-one should instantiate objects of this class directly) for azure core developers, it should not be showing up in intellisense right next to the credentials that we *do* want users of the library to use.

Reply:
@johanste Per our discussion, this is my understanding. https://github.com/Azure/azure-sdk-for-python/pull/28686.

And this is the end-user experience how to use the multi-key-AzureKeyCredential.

https://github.com/Azure/azure-sdk-for-python/pull/28686/files#diff-e1d74d0bf29d484a4d3388418079642e990d6a90a4d8bcf9ae259f2308cd2c72

Are we on same page?